### PR TITLE
Fix #8467: Prevent crash on long pressing an already selected card

### LINKED ISSUE #8467: [Bug] Long pressing a card below the end of the initial ListView crashes
###### Reproduction Steps

1. Go to card browser.
2. Long press on a card to select it.
3. After the card is selected, long press again.
4. Notice that the app crashes.

###### Expected Result

App should not crash.

###### Actual Result

The app crashes. Here are the logs:

```
java.lang.NullPointerException: Attempt to invoke virtual method 'android.view.View android.view.View.findViewById(int)' on a null object reference
        at com.ichi2.anki.CardBrowser.lambda$onCollectionLoaded$2$CardBrowser(CardBrowser.java:673)
        at com.ichi2.anki.-$$Lambda$CardBrowser$cpk2FDlr72riVUwAw9LYnTvoqho.onItemLongClick(Unknown Source:6)
        at android.widget.AbsListView.performLongPress(AbsListView.java:3288)
        at android.widget.AbsListView$CheckForLongPress.run(AbsListView.java:3205)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:201)
        at android.app.ActivityThread.main(ActivityThread.java:6864)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:873)
```

###### Research

- [x] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [x] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [x] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by ShridharGoel**
## Purpose / Description
App should not crash on long pressing an already selected card in the card browser.
 
## Fixes
Fixes #8467 

## Approach
`getChildAt()` method of `ListView` is wrongly used since it concerns the visible items and not the complete list of items. The approach followed to fix this is to make use of the `mCheckedCards` list. While iterating through the cards to select them, the `mCheckedCards` list is being updated and `onSelectionToggle()`  is being called to update the checked state of the list items.

## How Has This Been Tested?

Verified that the mentioned issue does not occur after this change.

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
This implies a bug somewhere else in the app - likely with selecting an invalid position

**Comment by david-allison-1** [comment]
Could you diagnose and fix/test the cause of the bug, rather than patching the result of the bug

**Comment by ShridharGoel** [comment]
Yes, I have found the issue. `getChildAt()` works with respect to the visible items in the ListView and not the actual positions in the list. Some alternative approach is needed here, which I'm trying to do.

**Comment by ShridharGoel** [comment]
If you have any suggestions, then please let me know.

**Comment by david-allison-1** [comment]
> Yes, I have found the issue. `getChildAt()` works with respect to the visible items in the ListView and not the actual positions in the list. Some alternative approach is needed here, which I'm trying to do.

I would expect that the problem is that the input to that function is bad, is this not the case?

**Comment by ShridharGoel** [comment]
Yes, the input to that function is wrong.

**Comment by david-allison-1** [comment]
Okay, so to fix this issue:

* work out why the input was bad, possibly use a debugger
* write a test which fails
* fix the code, and ensure the test passes

**Comment by ShridharGoel** [comment]
For confirmation regarding the expected behaviour, suppose I select the card at position 8 by long pressing. Then, I scroll and long press on the card at position 15. This should lead to the selection of all cards from position 8 to 15, right?

**Comment by david-allison-1** [comment]
> For confirmation regarding the expected behaviour, suppose I select the card at position 8 by long pressing. Then, I scroll and long press on the card at position 15. This should lead to the selection of all cards from position 8 to 15, right?

yep

**Comment by ShridharGoel** [comment]
Updated. Please have a look.

Regarding the tests, this requires an instrumentation test, right?

**Comment by madelesi** [comment]
This is closely related to #8460. Adding your null check prevents crashing in #8460, but does not toggle the checkbox for all the cards which need scrolling (which is expected).

@david-allison-1 
> I would expect that the problem is that the input to that function is bad, is this not the case?
Do you mean the ```getChildAt()``` function? Because the input to it seems correct to me

**Comment by david-allison-1** [comment]
> Updated. Please have a look.
> 
> Regarding the tests, this requires an instrumentation test, right?

Likely better under robolectric.

I'm on my phone, but I haven't seen a diagnosis for why the bug occurred, only a fix and that's a little concerning

**Comment by mikehardy** [comment]
I don't believe it will require an instrumentation test, RobolectricTest provides you with a collection where you may set up a data test fixture, and ability to start Activities and get View elements etc. You can see in CardTemplateEditorTest for instance doing some pretty heavy work with Activity lifecycles and checking view contents / altering them / asserting internal Activity state etc

**Comment by ShridharGoel** [comment]
> I'm on my phone, but I haven't seen a diagnosis for why the bug occurred, only a fix and that's a little concerning

I have updated the details under the `Approach` section of the PR description.

**Comment by ShridharGoel** [comment]
> I don't believe it will require an instrumentation test, RobolectricTest provides you with a collection where you may set up a data test fixture, and ability to start Activities and get View elements etc. You can see in CardTemplateEditorTest for instance doing some pretty heavy work with Activity lifecycles and checking view contents / altering them / asserting internal Activity state etc

Thank you.

**Comment by ShridharGoel** [comment]
> I don't believe it will require an instrumentation test, RobolectricTest provides you with a collection where you may set up a data test fixture, and ability to start Activities and get View elements etc. You can see in CardTemplateEditorTest for instance doing some pretty heavy work with Activity lifecycles and checking view contents / altering them / asserting internal Activity state etc

Okay, I'll try this and update. Thanks.

**Comment by ShridharGoel** [comment]
I have already tried creating this test in `CardBrowserTest`:

```
@Test
public void checkMultipleCardSelection() {
    CardBrowser browser = getBrowserWithNotes(10);
    selectOneOfManyCards(browser, 4);
    selectOneOfManyCards(browser, 8);
    assertThat(browser.checkedCardCount(), is(5));
}
```

Here, the card at position 4 is getting long clicked and then the one at position 8 (I have modified the `selectOneOfManyCards` method). But this test case is passing even when in actual the code leads to a crash. That's why I thought that an instrumentation test might be required. Any help on this? Does the above test look fine?

**Comment by david-allison-1** [comment]
You'll likely need to try emulating a long press in the test.

**Comment by madelesi** [comment]
> suppose I select the card at position 8 by long pressing. Then, I scroll and long press on the card at position 15. This should lead to the selection of all cards from position 8 to 15, right?

As a heads-up, the current work would select all cards from 9 to 15 in your example. (It toggles all cards in the range, meaning the first card which was checked will become unchecked). Setting i to ```Math.Min(mLastSelectedPosition, position)+1``` in the for-loop seems a good fix.

**Comment by ShridharGoel** [comment]
Detailed explanation:

Suppose the item with index 4 in the list is currently the first visible item on the screen. The method `getChildAt()` would require an input of 0 to refer to that item since it's at the 0th position (index) in the currently visible items, while in actual it's position (index) is 4 in the list. This was the reason of the bug, which this PR aims to solve. I have removed the dependency on `getChildAt()` method since it was not suitable for this use-case.

Also, because of this reason, it's not clear how to write a test which should fail for previous code and pass for this one.

**Comment by ShridharGoel** [comment]
> As a heads-up, the current work would select all cards from 9 to 15 in your example. (It toggles all cards in the range, meaning the first card which was checked will become unchecked). Setting i to `Math.Min(mLastSelectedPosition, position)+1` in the for-loop seems a good fix.

Thanks, I have updated the functionality according to the behaviour which seemed expected from the previous code.

According to the current functionality, multiple items are not being unchecked at once. Suppose the items from index 1 to index 10 are checked. If we long press on item with index 5, then only that will get unchecked. Then, if we long press on item at index 8, then again only that item will get unchecked.

This is the current behaviour. If this is not as intended, then do let me know.

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
I'll need to apologise here, I misunderstood the initial commit - the code didn't handle `mCheckedCards`, but the intuition was correct.

The fix will need commenting to explain why the code looks the way it does, and what is happening for both visible and invisible cards

----

I've taken a look at this issue in a debugger.

The issue is caused by: https://stackoverflow.com/a/6767006, and is triggered when there is more than one page of items in the CardBrowser

This means that to handle checking of unseen cards, we need to:

* Check the visible items on the screen, being aware that `getChildAt` returns the logical position, and not the physical position
* Modify our abstraction for checked cards: `mCheckedCards`
* Add a regression test

**Comment by ShridharGoel** [comment]
> The fix will need commenting to explain why the code looks the way it does, and what is happening for both visible and invisible cards

I'll update it with comments.

> The issue is caused by: https://stackoverflow.com/a/6767006, and is triggered when there is more than one page of items in the CardBrowser

That's correct and I had also taken reference from this StackOverflow answer.
More specifically, the crash happens after scrolling from the first screen of visible items.

> Check the visible items on the screen, being aware that getChildAt returns the logical position, and not the physical position

Using `getChildAt()` is not required, the required functionality can be handled with `mCheckedCards` as done in this PR.

> Modify our abstraction for checked cards: mCheckedCards

Is there anything wrong with `mCheckedCards`? It seems to be fine.

> Add a regression test

I'll add a unit test to ensure that the functionality is working.

**Comment by ShridharGoel** [comment]
> The fix will need commenting to explain why the code looks the way it does, and what is happening for both visible and invisible cards

I'll update it with comments.

> The issue is caused by: https://stackoverflow.com/a/6767006, and is triggered when there is more than one page of items in the CardBrowser

That's correct and I had also taken reference from this StackOverflow answer.
More specifically, the crash happens after scrolling from the first screen of visible items.

> Check the visible items on the screen, being aware that getChildAt returns the logical position, and not the physical position

Using `getChildAt()` is not required, the required functionality can be handled with `mCheckedCards` as done in this PR.

> Modify our abstraction for checked cards: mCheckedCards

Is there anything wrong with `mCheckedCards`? It seems to be fine.

> Add a regression test

I'll add a unit test to ensure that the functionality is working.

**Comment by ShridharGoel** [comment]
> The fix will need commenting to explain why the code looks the way it does, and what is happening for both visible and invisible cards

I will update it with more comments.

> The issue is caused by: https://stackoverflow.com/a/6767006, and is triggered when there is more than one page of items in the CardBrowser

That's correct and I had also taken reference from this StackOverflow answer.
More specifically, the crash happens after scrolling from the first screen of visible items.

> Check the visible items on the screen, being aware that getChildAt returns the logical position, and not the physical position

Using `getChildAt()` is not required, the required functionality can be handled with `mCheckedCards` as done in this PR.

> Modify our abstraction for checked cards: mCheckedCards

Is there anything wrong with `mCheckedCards`? It seems to be fine.

> Add a regression test

I'll add a unit test to ensure that the functionality is working.

**Comment by ShridharGoel** [comment]
> The fix will need commenting to explain why the code looks the way it does, and what is happening for both visible and invisible cards

I will update it with more comments.

> The issue is caused by: https://stackoverflow.com/a/6767006, and is triggered when there is more than one page of items in the CardBrowser

That's correct and I had also taken reference from this StackOverflow answer.
More specifically, the crash happens after scrolling from the first screen of visible items.

> Check the visible items on the screen, being aware that getChildAt returns the logical position, and not the physical position

Using `getChildAt()` is not required, the required functionality can be handled with `mCheckedCards` as done in this PR.

> Modify our abstraction for checked cards: mCheckedCards

Is there anything wrong with `mCheckedCards`? It seems to be fine.

> Add a regression test

I'll add a unit test.

**Comment by ShridharGoel** [comment]
@david-allison-1 Please have a look at the unit test.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 1]
```diff
@@ -527,6 +527,14 @@ public void checkSearchString() {
         assertThat("Results should only be from the selected deck", cardBrowser.getCardCount(), is(1));
     }
 
+    @Test
+    public void checkIfLongSelectChecksAllCardsInBetween() {
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:531]: This test passes before the change was made

**Comment by david-allison-1** [comment]
> > Check the visible items on the screen, being aware that getChildAt returns the logical position, and not the physical position
> 
> Using `getChildAt()` is not required, the required functionality can be handled with `mCheckedCards` as done in this PR.

This PR doesn't seem correct - 

* select 0
* tap 3
* long tap 3

I'd expect (0,1,2,3) to be selected

> > Modify our abstraction for checked cards: mCheckedCards
> 
> Is there anything wrong with `mCheckedCards`? It seems to be fine.

I was unclear: how we check cards inside and outside the visible view range will need to be better abstracted.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 2]
```diff
@@ -98,22 +98,22 @@ public void selectNoneIsVisibleOnceSelectAllCalled() {
     public void selectNoneIsVisibleWhenSelectingOne() {
         CardBrowser browser = getBrowserWithMultipleNotes();
         advanceRobolectricLooperWithSleep();
-        selectOneOfManyCards(browser);
+        selectOneOfManyCards(browser, 0);
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:101]: Design: Could we add a method with a single parameter which defaults to 0

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 3]
```diff
@@ -565,11 +573,10 @@ private void deleteCardAtPosition(CardBrowser browser, int positionToCorrupt) {
         browser.clearCardData(positionToCorrupt);
     }
 
-    private void selectOneOfManyCards(CardBrowser browser) {
+    private void selectOneOfManyCards(CardBrowser browser, int position) {
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:576]: Design: This method now performs more than selection. Could we add in another method for multiple selection, and cause `selectOneOfManyCards` to fail if it invokes the new multi-select code.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 4]
```diff
@@ -667,27 +667,30 @@ public void onNothingSelected(AdapterView<?> parent) {
...
-                    if (!cb.isChecked()) {
-                        cb.toggle();
-                        onCheck(i, childView);
+                if (!mCheckedCards.contains(mCardsListView.getItemAtPosition(position))) {
+                    for (int i = Math.min(mLastSelectedPosition, position); i <= Math.max(mLastSelectedPosition, position); i++) {
+                        // Fetch card at position i.
+                        CardCache card = (CardCache) mCardsListView.getItemAtPosition(i);
+
+                        // If card is not in the list of checked cards, then add it.
+                        if (!mCheckedCards.contains(card)) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:676]:
```suggestion
                        if (mCheckedCards.add(card)) {
```

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 5]
```diff
@@ -667,27 +667,30 @@ public void onNothingSelected(AdapterView<?> parent) {
...
-                cb.toggle();
-                onCheck(position, view);
-                recenterListView(view);
-                mCardsAdapter.notifyDataSetChanged();
             }
+
+            // click on whole cell triggers select
+            CheckBox cb = view.findViewById(R.id.card_checkbox);
+            cb.toggle();
+            onCheck(position, view);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:691]: Doesn't this call `onCheck` twice?

**Comment by ShridharGoel** [comment]
> * select 0
> * tap 3
> * long tap 3
> 
> I'd expect (0,1,2,3) to be selected

The requirement was not perfectly clear to me earlier. I have updated the code now.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java (hunk 5 above)
Comment (ShridharGoel) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java]: It was outside the if-else block and was being called just once I think. But now the implementation has changed in the updated code.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 1 above)
Comment (ShridharGoel) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:531]: The issue with the previous code was that wrong index was being passed as input to `getChildAt()`. I think that cannot be replicated with Roboelectric unless we explicitly pass a wrong index.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 2 above)
Comment (ShridharGoel) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java]: Updated.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 1 above)
Comment (david-allison-1) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:531]: https://github.com/ankidroid/Anki-Android/blob/34c9f25038d24e3bf9f338fa307b1bcb3610b5ba/AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java#L578-L579

`toSelect.getChildCount()` is 9 here.

We need to fix the test (a null view is passed through currently), but the bug is reproducible in a unit test.

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 6]
```diff
@@ -665,15 +665,32 @@ public void onNothingSelected(AdapterView<?> parent) {
...
-                    CheckBox cb = childView.findViewById(R.id.card_checkbox);
-                    if (!cb.isChecked()) {
-                        cb.toggle();
-                        onCheck(i, childView);
+                    // Fetch card at position i.
+                    CardCache card = (CardCache) mCardsListView.getItemAtPosition(i);
+
+                    // If card is not in the list of checked cards, then add it.
+                    if (mCheckedCards.add(card)) {
+                        onSelectionChanged();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:693]: This is extremely inefficient if we select many cards, keep a variable, and perform the action at the end

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 7]
```diff
@@ -1966,7 +1972,7 @@ public void actualOnPostExecute(@NonNull CardBrowser browser, List<CardCache> re
         private void handleSearchResult() {
             Timber.i("CardBrowser:: Completed doInBackgroundSearchCards Successfully");
             updateList();
-            
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:1975]: Could you revert this spacing change

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 1 above)
Comment (david-allison-1) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:531]: Could you ensure the test fails before you make the change to fix the bug

**Comment by mikehardy** [comment]
Superceded by #8665
